chain/memdb/store.go

```diff
@@ -0,0 +1,195 @@
+package memdb
+
+import (
+	"context"
+	"io"
+	"sort"
+	"sync"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/chain/errors"
+)
+
+// Store represents access to the in-memory storage for beacon management.
+type Store struct {
+	storeMtx *sync.RWMutex
+	store    []*chain.Beacon
+	maxSize  int
+}
+
+// NewStore returns a new store that provides the CRUD based API needed for
+// supporting drand serialization.
+func NewStore() *Store {
+	const maxSize = 10
+
+	return &Store{
+		storeMtx: &sync.RWMutex{},
+		store:    []*chain.Beacon{},
+		maxSize:  maxSize,
+	}
+}
+
+func (m *Store) Len(_ context.Context) (int, error) {
+	m.storeMtx.RLock()
+	defer m.storeMtx.RUnlock()
+
+	return len(m.store), nil
+}
+
+func (m *Store) Put(_ context.Context, beacon *chain.Beacon) error {
+	m.storeMtx.Lock()
+	defer m.storeMtx.Unlock()
+	defer func() {
+		if len(m.store) > m.maxSize {
+			m.store = m.store[len(m.store)-m.maxSize:]
+		}
+	}()
+
+	found := false
+	defer func() {
+		sort.Slice(m.store, func(i, j int) bool {
+			return m.store[i].Round < m.store[j].Round
+		})
+	}()
+
+	for _, sb := range m.store {
+		if sb.Round == beacon.Round {
+			found = true
+			break
+		}
+	}
+
+	if found {
+		return nil
+	}
+
+	m.store = append(m.store, beacon)
+	return nil
+}
+
+func (m *Store) Last(_ context.Context) (*chain.Beacon, error) {
+	m.storeMtx.RLock()
+	defer m.storeMtx.RUnlock()
+
+	if len(m.store) == 0 {
+		return nil, errors.ErrNoBeaconStored
+	}
+
+	result := m.store[len(m.store)-1]
+	return result, nil
+}
+
+func (m *Store) Get(_ context.Context, round uint64) (*chain.Beacon, error) {
+	m.storeMtx.RLock()
+	defer m.storeMtx.RUnlock()
+
+	for _, beacon := range m.store {
+		if beacon.Round == round {
+			return beacon, nil
+		}
+	}
+
+	return nil, errors.ErrNoBeaconStored
+}
+
+func (m *Store) Cursor(ctx context.Context, f func(context.Context, chain.Cursor) error) error {
```

Comment:
very minor, but using `m` rather than `s` confused my brain for a moment

Reply:
I'll update it.